README.md
# dicekit-fp

A functional-programming-friendly library for rolling virtual dice. It provides pure-functional way to create random numbers and dice (functions that create random integers), combine and modiify the dice results, and convert strings like "2d6+2" to functions. It has no dependencies and is about 2KB minified.

Instead of using objects, the library creates random number generators based on the parameters you provide. The library uses higher-order functions to modify the results in order to modify or multiply the results of the dice rolls. Most of the functions return a `() => number` (aliased in the Typescript code as `NumberGenerator`).

The functions that produce random numbers, such as `createDie()`, typicially have two versions: one that takes a random number generator (RNG) as the first parameter and have the suffix `With` as in `createDieWith()`, the other which is already seeded with `Math.random()` as the RNG as in `createDie()`

## Use

### Create a basic die with `createDie()`

```typescript
// takes a number of faces for the die as an argument and returns a function that generates ints.
const d6 = createDie(6);

// random int between 1 and 6
d6();
```

You can use any number (>=2) of faces

```typescript
const coinFlip = createDie(2);
const devilDie = createDie(666);
```

### Create multiple dice with a modifier in one step using `createDice()`

```typescript
// takes a number of faces, number of dice (default 1), and modifier (default +0)
// 2d6+3
const _2d6plus3 = createDice(6, 2, 3);

// Random result of rolling 2 six sided dice and adding 3 to the result.
// (1..6) + (1..6) + 3
_2d6plus3();
// Note, this is not the same as a random int between 5..15.
// Because each die gets rolled independently, the results are more likely to follow a normal distribution curve favoring the middle numbers (like 7) more than the edge numbers (like 2 and 12)

// 2d6+0
const _2d6 = createDice(6, 2);
// 1d6+0
const _1d6 = createDice(6);
```

### Combine multiple dice functions with `combineDice()`

```typescript
const d6 = createDie(6);
const d8 = createDie(8);
// always returns 2
const two = () => 2;

// 2d6+1d8+2
const _2d6_1d8_2 = combineDice([d6, d6, d8, two]);

// generate random int
// (1..6) + (1..6) + (1..8) + 2
_2d6_1d8_2();
```

### Dice Strings

Using `parseDiceString(s)` will convert a string into a function that rolls dice based on the input. The format of the string is inspired by Tabletop RPG dice notation.

`${multiplier}d${sides}+${modifier}`

Where:

- `multiplier` is the number of dice (default is 1)
- Use `d` or `D` to separate the multiplier from the number of sides on the dice.
- `sides` is the number of sides on the dice. The range will be `1` to `sides`, inclusive.
- `modifier` is an integer to add to the total

**Examples**

- "d6" -> roll one 6-sided-die`
- "2d6" -> roll two 6-sided-dice and sum results`
- "3d6+2" -> roll three 6-sided-dice, add 2 to the sum.`
- "2d12-2" -> roll two 12-sided-dice, subtract 2 from the sum.`
- "1d12-5+2d20+4+d4" -> roll one 12-sided-die, two 20-sided-dice, and one 4-sided-die, subtract 5 and add 4 to the sum.`

Whitespace and letter case are ignored.

### Custom dice with `createCustomDie()`

To create a die that isn't typical, you can use `createCustomDie()` and provide an array of dice face values.

```typescript
// Generate a random color string
const colors = ["red", "orange", "yellow", "green", "blue", "indigo", "violet"];
const colorDie = createCustomDie(colors);
colorDie(); // "blue"

// Create a die that lands on 6 more often.
const weightedD6 = createCustomDie([1, 2, 3, 4, 5, 6, 6, 6]);
weightedD6(); // 6, probably

// If the values are numerical, they can be combined with other dice functions...
const weightedDice = combineDice([weightedDie, weightedDie]);
```

### Custom Random seed

All the fuctions that rely on a random number generator come with a version that uses `Math.random` for the RNG. To use a different generator, use the alternate versions with the suffix `With` as in `createDieWith(f)` where `f` is a function that takes no arguments and returns a number, `n`, where `0.0<=n<1.0`

```typescript
const myRNG = randomSeed.create("my seed");
const createDieSeeded = createDieWith(myRNG);

const d6 = createDieSeeded(6);
d6(); // uses myRNG for generating numbers.
```

### Other functions

TBD. Meanwhile, please review `test` folder for examples.

_Below this line is boilerplate from the template_

=====

# Vite npm package template

A template for an npm package using:

- vite
- typescript
- jest
- prettier & eslint
- controlled commits with commitizen, lint-staged, etc.
- & more!

Inspired by this post by [Onur Önder](https://onderonur.netlify.app/blog/creating-a-typescript-library-with-vite/)

## Setup

1. Install packages.
1. Edit the top of the package.json file to include the correct package name and description. You can find and replace for "\<LIBRARY\>"
1. Edit the README.md
1. You may need to review the `vite.config.js` depending on your library structure.

## Scripts

- `lint` : lint the code. You can also do `fix` to auto-fix.
- `test` : test with jest. Also `test:coverage` and `test:watch`
- `build` : bundle the code
- `commit` & `release` : see below
- `pushpub`: Used after `release` it pushes and publishes the library.
- `deploy`: Build the demo code (index.html) and publish to gh-pages

## Commits & Releases

Code is automatically linted before being committed. I recommend installing the plugins for eslint and prettier in your code editor. You can attempt to fix linting issues with `yarn fix`.

When ready to commit, please commit using `yarn commit` to use commitizen for standard format commits.

When ready to release use `yarn release` with the `-r patch|minor|major` flag (default without the flag is `patch`).

You'll then need to publish your changes separately. That can be done with `yarn pushpub`.

You can deploy an updated demo to github pages using `yarn deploy`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.4.0](https://github.com/mimshwright/dicekit/compare/v2.3.0...v2.4.0) (2023-02-15)

### Features

- **simplifydicestring:** added the ability to convert a dice string into a simplified version ([437781d](https://github.com/mimshwright/dicekit/commit/437781d88bf7fe5c9904cf5ea9a43abc68ce0a22))

## [2.3.0](https://github.com/mimshwright/dicekit/compare/v2.2.1...v2.3.0) (2023-02-14)

### Features

- **tokentostring:** made the tokenize function public and added a way to convert back to strings ([68524c9](https://github.com/mimshwright/dicekit/commit/68524c9c00e85b90fcc3ba2f0e5c34786e4ff056))

### [2.2.1](https://github.com/mimshwright/dicekit/compare/v2.2.0...v2.2.1) (2023-02-09)

## [2.2.0](https://github.com/mimshwright/dicekit/compare/v2.1.0...v2.2.0) (2023-02-09)

### Features

- **range:** added getRange() ([8e3fc4f](https://github.com/mimshwright/dicekit/commit/8e3fc4fd73c95f793000b2ee5e131bf6704bfa6b))

## [2.1.0](https://github.com/mimshwright/dicekit/compare/v2.0.1...v2.1.0) (2023-02-07)

### Bug Fixes

- **build:** fixed major issue in build where the package didn't work ([1577f35](https://github.com/mimshwright/dicekit/commit/1577f3537497cc5824114a73043e7c5f2241ecc1))

### [2.0.1](https://github.com/mimshwright/dicekit/compare/v2.0.0...v2.0.1) (2023-02-07)

### Bug Fixes

- **cleanup:** removed old package file ([1333ea5](https://github.com/mimshwright/dicekit/commit/1333ea50a499e5d1595b306d5ee1374356d7fdf8))

## [2.0.0](https://github.com/mimshwright/dicekit/compare/v1.0.4...v2.0.0) (2023-02-07)

### ⚠ BREAKING CHANGES

- **vite:** Entirely new build system but API should remain the same.

- **vite:** converted project to a vite project ([7f2fc28](https://github.com/mimshwright/dicekit/commit/7f2fc28b63196f4edda749bfc6202ae3236dd067))

### [0.2.1](https://github.com/mimshwright/<LIBRARY>/compare/v0.2.0...v0.2.1) (2023-02-04)

## [0.2.0](https://github.com/mimshwright/<LIBRARY>/compare/v0.1.2...v0.2.0) (2023-02-04)

### [0.1.2](https://github.com/mimshwright/<LIBRARY>/compare/v0.1.1...v0.1.2) (2022-12-04)

### 0.1.1 (2022-12-04)

# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.


test/blah.test.ts
import {
  identity,
  addToRoll,
  callMultipleTimes,
  random,
  randomBoolean,
  randomIntegerBetween,
  randomInteger,
  createDie,
  createDice,
  multipleDice,
  combineDice,
  parseDiceString,
} from "../src/index";

const max = (nums: number[]) =>
  nums.reduce((a, b) => Math.max(a, b), Number.NEGATIVE_INFINITY);

const min = (nums: number[]) =>
  nums.reduce((a, b) => Math.min(a, b), Number.POSITIVE_INFINITY);

const testRoll = callMultipleTimes;

const testRollSm = testRoll(200);
const testRollMed = testRoll(1000);
const testRollLrg = testRoll(5000);
const testRollXLrg = testRoll(100000);

// Test the test utils.

describe("test utils", () => {
  describe("identity()", () => {
    it("Takes an argument, a, returns a function that returns a (ignoring any arguments provided to it). ", () => {
      expect(identity(1)()).toBe(1);
      expect(identity("foo")("bar")).toBe("foo");
    });
  });

  describe("callMultipleTimes()", () => {
    let x = 0;
    const testFunc = () => {
      x += 1;
    };

    callMultipleTimes(10)(testFunc);

    expect(x).toBe(10);
  });

  describe("min()", () => {
    it("Returns the minimum value in a number[].", () => {
      expect(min([9, 50, 7, 1, -13, 102])).toBe(-13);
    });
  });

  describe("max()", () => {
    it("Returns the maximum value in a number[].", () => {
      expect(max([9, 50, 7, 1, -13, 102])).toBe(102);
    });
  });
});

// Test the main library.

describe("dicekit", () => {
  describe("[WARNING: In rare cases, these tests can fail due to random values. Please test again if this happens or raise the number of testRolls!]", () => {
    describe("random()", () => {
      it("Returns a value between 0 and 1", () => {
        const results = testRollMed(random);

        expect(min(results)).toBeGreaterThan(0);
        expect(max(results)).toBeLessThan(1);
      });
    });

    describe("randomIntegerBetween()", () => {
      it("Creates random numbers in a range.", () => {
        const results = testRollMed(() => randomIntegerBetween(50, 10));

        expect(min(results)).toBe(10);
        expect(max(results)).toBe(50);
      });

      it("Low value defaults to 0.", () => {
        const results = testRollMed(() => randomIntegerBetween(10));

        expect(min(results)).toBe(0);
        expect(max(results)).toBe(10);
      });

      it("Negative numbers work.", () => {
        const results = testRollMed(() => randomIntegerBetween(-10, -50));

        expect(max(results)).toBe(-10);
        expect(min(results)).toBe(-50);
      });

      it("Range is inclusive.", () => {
        const results = testRollSm(() => randomIntegerBetween(2, 1));

        expect(min(results)).toBe(1);
        expect(max(results)).toBe(2);
      });

      it("Expects max to be the first parameter and min to be the second, but if min > max, it switches them automatically.", () => {
        const resultsIncorrect = testRollSm(() => randomIntegerBetween(1, 3));
        const resultsCorrect = testRollSm(() => randomIntegerBetween(3, 1));

        expect(min(resultsCorrect)).toBe(1);
        expect(min(resultsIncorrect)).toBe(1);
        expect(max(resultsCorrect)).toBe(3);
        expect(max(resultsIncorrect)).toBe(3);
      });
    });

    describe("randomInteger()", () => {
      it("Is a unary form of randomIntegerBetween(). Only takes a maximum and assumes minimum is 0.", () => {
        const results = testRollMed(() => randomInteger(10));
        const resultsBetween = testRollMed(() => randomIntegerBetween(10, 0));

        expect(min(results)).toBe(min(resultsBetween));
        expect(max(results)).toBe(max(resultsBetween));
      });
    });
    describe("randomBoolean()", () => {
      it("Is syntactic sugar for randomIntegerBetween(0,1) coerced to a boolean value", () => {
        const results = testRoll(100)(randomBoolean);
        const trueCount = results.filter((b) => b === true).length;
        const falseCount = results.filter((b) => b === false).length;

        expect(trueCount).toBeGreaterThan(35);
        expect(falseCount).toBe(100 - trueCount);
      });
    });

    describe("createDie()", () => {
      it("Returns a function that generates random numbers between 1 and 'sides'", () => {
        const d6 = createDie(6);
        const results = testRollSm(d6);

        expect(d6).toBeInstanceOf(Function);
        expect(min(results)).toBe(1);
        expect(max(results)).toBe(6);
      });
      it("Range is inclusive'", () => {
        const coinFlip = createDie(2);
        const results = testRollSm(coinFlip);

        expect(min(results)).toBe(1);
        expect(max(results)).toBe(2);
      });
    });

    describe("multipleDice()", () => {
      it("Takes a die function and returns a function that calls it multiple times.", () => {
        const d6 = createDie(6);
        const multipleD6 = multipleDice(d6);
        const d6x3 = multipleD6(3);
        const results = testRollLrg(d6x3);

        expect(multipleD6).toBeInstanceOf(Function);
        expect(d6x3).toBeInstanceOf(Function);

        expect(min(results)).toBe(3);
        expect(max(results)).toBe(18);
        // make sure it's generating numbers in between
        expect(results.indexOf(4)).toBeGreaterThanOrEqual(0);

        const coin = createDie(2);
        const coinx10 = multipleDice(coin)(10);
        const flips = testRollLrg(coinx10);

        expect(min(flips)).toBe(10);
        expect(max(flips)).toBe(20);
      });
    });

    describe("constant()", () => {
      it("Takes a number value and returns a function that always returns that number. K-combinator", () => {
        const one = identity(1);

        expect(one).toBeInstanceOf(Function);
        expect(one()).toBe(1);
        expect(identity(0)()).toBe(0);
        expect(identity(999)()).toBe(999);
      });
    });

    describe("addToRoll()", () => {
      it("Takes a function that returns a number. When called, it calls the function and adds a constant value to the result.", () => {
        const d6 = createDie(6);
        const plus2 = addToRoll(2);
        const d6plus2 = plus2(d6);
        const results = testRollLrg(d6plus2);
        expect(min(results)).toBe(3);
        expect(max(results)).toBe(8);

        const multipleD6 = multipleDice(d6);
        const d6x3plus2 = plus2(multipleD6(3));
        const resultsMultiple = testRollLrg(d6x3plus2);
        expect(min(resultsMultiple)).toBe(5);
        expect(max(resultsMultiple)).toBe(20);
      });
    });

    describe("createDice()", () => {
      describe("Creates multiple dice (all of the same number of sides) and adds a modifier to the result.", () => {
        it("Creates several dice and adds a fixed number to the output.", () => {
          const _2d6plus4 = createDice(6, 2, 4);

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });
        it("Uses 0 for the default modifier.", () => {
          const _2d6 = createDice(6, 2);

          const result = testRollLrg(_2d6);
          expect(min(result)).toBe(2 * 1 + 0);
          expect(max(result)).toBe(2 * 6 + 0);
        });
        it("Uses 1 for the default multiplier.", () => {
          const _d6 = createDice(6);
          const result = testRollLrg(_d6);
          expect(min(result)).toBe(1 * 1 + 0);
          expect(max(result)).toBe(1 * 6 + 0);
        });
      });
    });

    describe("combineDice()", () => {
      describe("Takes an array of dice roll functions and returns one that combines the results of all.", () => {
        it("Is a function that returns a function.", () => {
          const one = identity(1);
          const combined = combineDice([one]);

          expect(combineDice).toBeInstanceOf(Function);
          expect(combined).toBeInstanceOf(Function);
          expect(combined()).toBe(1);
        });

        it("Array must not be empty", () => {
          expect(() => {
            combineDice([]);
          }).toThrow();
        });

        it("Combines dice functions. Functions must be in the form `() => number`", () => {
          const d6 = createDie(6);
          const plus2 = identity(2);

          const _2d6plus2 = combineDice([d6, d6, plus2]);
          expect(_2d6plus2).toBeInstanceOf(Function);

          const results = testRollMed(_2d6plus2);
          expect(min(results)).toBe(4);
          expect(max(results)).toBe(14);

          const d8 = createDie(8);
          const d12 = createDie(12);
          const _2d12 = multipleDice(d12)(2);
          const _2d12plus3 = addToRoll(3)(_2d12);

          const _2d12plus3plus1d8minus5 = combineDice([
            _2d12plus3,
            d8,
            identity(-5),
          ]);

          const results2 = testRollXLrg(_2d12plus3plus1d8minus5);
          expect(min(results2)).toBe(2 + 3 + 1 - 5);
          expect(max(results2)).toBe(2 * 12 + 3 + 8 - 5);
        });
      });
    });

    describe("parseDiceString()", () => {
      const p = parseDiceString;
      describe("Takes strings in the format `NdS+K` and converts to a dice rolling function. e.g. 2d6+4 => Roll 2x 6-sided dice and add +4 to the result. ", () => {
        it("Basic function", () => {
          const _2d6plus4 = p("2d6+4");

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });

        it("Whitespace ignored.", () => {
          const _2d6plus4 = p(`2 d6    + 
          4`);

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });

        it("Case insensitive.", () => {
          const _2d6plus4 = p(`2D6+4`);

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });

        it("Negative modifiers work as expected", () => {
          const _3d6minus2 = p("3d6-2");

          const result = testRollLrg(_3d6minus2);
          expect(min(result)).toBe(3 * 1 - 2);
          expect(max(result)).toBe(3 * 6 - 2);
        });
        it("Multiple modifiers are all summed", () => {
          const _2d6plus4 = p("2d6+2+1+1");

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });
        it("Modifiers can come before the dice.", () => {
          const _2d6plus4 = p("4+2d6");

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });

        it("Dice can be split up.", () => {
          const _2d6plus4 = p("1d6+2+1d6+2");

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });

        it("Multiplier assumed to be 1 if not specified", () => {
          const _d6plus2 = p("d6+2");

          const result = testRollLrg(_d6plus2);
          expect(min(result)).toBe(1 * 1 + 2);
          expect(max(result)).toBe(1 * 6 + 2);
        });

        it("Modifier optional.", () => {
          const _2d6 = p("2d6");
          const result = testRollLrg(_2d6);
          expect(min(result)).toBe(2 * 1);
          expect(max(result)).toBe(2 * 6);

          const _2d6plus2d12 = p("2d6+2d12");
          const result2 = testRollXLrg(_2d6plus2d12);
          expect(min(result2)).toBe(2 * 1 + 2 * 1);
          expect(max(result2)).toBe(2 * 6 + 2 * 12);
        });
      });
    });
  });
});


test/index.test.ts
import {
  DiceToken,
  NumberGenerator,
  Modifier,
  DiceTokensWithModifier,
} from "./../src/types";
import { describe, it, expect } from "vitest";

import {
  identity,
  testRollSm,
  testRollMed,
  testRollLrg,
  testRollXLrg,
  min,
  max,
  count,
} from "../src/utils";
import {
  createDieWith,
  createDie,
  createCustomDieWith,
  createCustomDie,
  createDiceWith,
  createDice,
  parseDiceStringWith,
  parseDiceString,
  multipleDice,
  combineDice,
  addToRoll,
  getRange,
  tokensToString,
  tokenizeDiceString,
  simplifyDiceString,
} from "../src/index";

describe("dicekit", () => {
  const tbf = (f: unknown) => expect(f).toBeInstanceOf(Function);
  describe("createDieWith() &c.", () => {
    describe("Creates a die with your choice of RNG.", () => {
      it("Every random function has a version that allows you to specify the RNG.", () => {
        tbf(createDieWith);
        tbf(createDiceWith);
        tbf(createCustomDieWith);
        tbf(parseDiceStringWith);
      });
      it("Versions that use Math.random by default are also exported.", () => {
        tbf(createDie);
        tbf(createDice);
        tbf(createCustomDie);
        tbf(parseDiceString);
      });

      it("Can take any function that returns numbers, n, where 0.0>=n<1.0 ", () => {
        const sequence = [0.0, 0.2, 0.4, 0.6, 0.8, 0.99999999];
        let i = 0;
        const sequencedNumberGenerator = () => {
          const result = sequence[i];
          i = (i + 1) % sequence.length;
          return result;
        };

        const die = createDieWith(sequencedNumberGenerator)(6);
        expect(die()).toBe(1);
        expect(die()).toBe(2);
        expect(die()).toBe(3);
        expect(die()).toBe(4);
        expect(die()).toBe(5);
        expect(die()).toBe(6);
        expect(die()).toBe(1);
      });
    });
  });

  describe("[WARNING: In rare cases, these tests can fail due to random values. Please test again if this happens or raise the number of testRolls!]", () => {
    describe("createDie()", () => {
      it("Returns a function that generates random numbers between 1 and 'sides'", () => {
        const d6 = createDie(6);
        const results = testRollSm(d6);

        tbf(d6);
        expect(min(results)).toBe(1);
        expect(max(results)).toBe(6);
      });
      it("Range is inclusive'", () => {
        const coinFlip = createDie(2);
        const results = testRollSm(coinFlip);

        expect(min(results)).toBe(1);
        expect(max(results)).toBe(2);
      });
    });

    describe("createCustomDie()", () => {
      it("Returns a generator that maps each side to a different value.", () => {
        const weightedDie: NumberGenerator = createCustomDie([
          1, 2, 2, 3, 3, 3, 3, 4, 4, 5,
        ]);
        const results = testRollMed(weightedDie);
        const count1 = count(1)(results);
        const count3 = count(3)(results);

        expect(min(results)).toBe(1);
        expect(max(results)).toBe(5);
        expect(count3 / count1).toBeGreaterThan(3);
      });
      it("Can return a NumberGenerator or another type of generator.", () => {
        const colors = [
          "red",
          "orange",
          "yellow",
          "green",
          "blue",
          "indigo",
          "violet",
        ];
        const colorDie = createCustomDie(colors);
        const results = testRollSm(colorDie);

        expect(results[0]).toHaveProperty("length");
        colors.forEach((c) => expect(count(c)(results)).toBeGreaterThan(10));

        expect(createCustomDie([new Date()])()).toBeInstanceOf(Date);
      });

      it("Can still be combined with other functions if it's a number generator", () => {
        const ones = createDie(10);
        const tens = createCustomDie([10, 20, 30, 40, 50, 60, 70, 80, 90, 0]);

        const hundredSided = combineDice([ones, tens]);

        const results = testRollLrg(hundredSided);
        expect(min(results)).toBe(1);
        expect(max(results)).toBe(100);
      });
    });

    describe("multipleDice()", () => {
      it("Takes a die function and returns a function that calls it multiple times.", () => {
        const d6 = createDie(6);
        const multipleD6 = multipleDice(d6);
        const d6x3 = multipleD6(3);
        const results = testRollLrg(d6x3);

        tbf(multipleD6);
        tbf(d6x3);

        expect(min(results)).toBe(3);
        expect(max(results)).toBe(18);
        // make sure it's generating numbers in between
        expect(results.indexOf(4)).toBeGreaterThanOrEqual(0);

        const coin = createDie(2);
        const coinx10 = multipleDice(coin)(10);
        const flips = testRollLrg(coinx10);

        expect(min(flips)).toBe(10);
        expect(max(flips)).toBe(20);
      });
    });

    describe("addToRoll()", () => {
      it("Takes a function that returns a number. When called, it calls the function and adds a constant value to the result.", () => {
        const d6 = createDie(6);
        const plus2 = addToRoll(2);
        const d6plus2 = plus2(d6);
        const results = testRollLrg(d6plus2);
        expect(min(results)).toBe(3);
        expect(max(results)).toBe(8);

        const multipleD6 = multipleDice(d6);
        const d6x3plus2 = plus2(multipleD6(3));
        const resultsMultiple = testRollLrg(d6x3plus2);
        expect(min(resultsMultiple)).toBe(5);
        expect(max(resultsMultiple)).toBe(20);
      });
    });

    describe("createDice()", () => {
      describe("Creates multiple dice (all of the same number of sides) and adds a modifier to the result.", () => {
        it("Creates several dice and adds a fixed number to the output.", () => {
          const _2d6plus4 = createDice(6, 2, 4);

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });
        it("Uses 0 for the default modifier.", () => {
          const _2d6 = createDice(6, 2);

          const result = testRollLrg(_2d6);
          expect(min(result)).toBe(2 * 1 + 0);
          expect(max(result)).toBe(2 * 6 + 0);
        });
        it("Uses 1 for the default multiplier.", () => {
          const _d6 = createDice(6);
          const result = testRollLrg(_d6);
          expect(min(result)).toBe(1 * 1 + 0);
          expect(max(result)).toBe(1 * 6 + 0);
        });
      });
    });

    describe("combineDice()", () => {
      describe("Takes an array of dice roll functions and returns one that combines the results of all.", () => {
        it("Is a function that returns a function.", () => {
          const one = identity(1);
          const combined = combineDice([one]);

          tbf(combineDice);
          tbf(combined);
          expect(combined()).toBe(1);
        });

        it("Will throw an error if array is empty", () => {
          expect(() => {
            combineDice([]);
          }).toThrow();
        });

        it("Combines dice functions. Functions must be in the form `() => number`", () => {
          const d6 = createDie(6);
          const plus2 = identity(2);

          const _2d6plus2 = combineDice([d6, d6, plus2]);
          tbf(_2d6plus2);

          const results = testRollMed(_2d6plus2);
          expect(min(results)).toBe(4);
          expect(max(results)).toBe(14);

          const d8 = createDie(8);
          const d12 = createDie(12);
          const _2d12 = multipleDice(d12)(2);
          const _2d12plus3 = addToRoll(3)(_2d12);

          const _2d12plus3plus1d8minus5 = combineDice([
            _2d12plus3,
            d8,
            identity(-5),
          ]);

          const results2 = testRollXLrg(_2d12plus3plus1d8minus5);
          expect(min(results2)).toBe(2 + 3 + 1 - 5);
          expect(max(results2)).toBe(2 * 12 + 3 + 8 - 5);
        });
      });
    });

    describe("parseDiceString()", () => {
      const p = parseDiceString;
      describe("Takes strings in the format `NdS+K` and converts to a dice rolling function. e.g. 2d6+4 => Roll 2x 6-sided dice and add +4 to the result. ", () => {
        it("Basic function", () => {
          const _2d6plus4 = p("2d6+4");

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });

        it("Whitespace ignored.", () => {
          const _2d6plus4 = p(`2 d6    + 
          4`);

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });

        it("Case insensitive.", () => {
          const _2d6plus4 = p(`2D6+4`);

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });

        it("Negative modifiers work as expected", () => {
          const _3d6minus2 = p("3d6-2");

          const result = testRollLrg(_3d6minus2);
          expect(min(result)).toBe(3 * 1 - 2);
          expect(max(result)).toBe(3 * 6 - 2);
        });
        it("Multiple modifiers are all summed", () => {
          const _2d6plus4 = p("2d6+2+1+1");

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });
        it("Modifiers can come before the dice.", () => {
          const _2d6plus4 = p("4+2d6");

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });

        it("Dice can be split up.", () => {
          const _2d6plus4 = p("1d6+2+1d6+2");

          const result = testRollLrg(_2d6plus4);
          expect(min(result)).toBe(2 * 1 + 4);
          expect(max(result)).toBe(2 * 6 + 4);
        });

        it("Multiplier assumed to be 1 if not specified", () => {
          const _d6plus2 = p("d6+2");

          const result = testRollLrg(_d6plus2);
          expect(min(result)).toBe(1 * 1 + 2);
          expect(max(result)).toBe(1 * 6 + 2);
        });

        it("Modifier optional.", () => {
          const _2d6 = p("2d6");
          const result = testRollLrg(_2d6);
          expect(min(result)).toBe(2 * 1);
          expect(max(result)).toBe(2 * 6);

          const _2d6plus2d12 = p("2d6+2d12");
          const result2 = testRollXLrg(_2d6plus2d12);
          expect(min(result2)).toBe(2 * 1 + 2 * 1);
          expect(max(result2)).toBe(2 * 6 + 2 * 12);
        });

        it("Should fail if you provide a number of dice which are < 0", () => {
          const _neg2d6 = p("-2d6");
          const result = testRollLrg(_neg2d6);

          // todo: add more helpful error.

          expect(min(result)).toBe(NaN);
          expect(max(result)).toBe(NaN);
        });
      });
    });

    describe("tokenizeDiceString()", () => {
      it("Converts strings to tokens", () => {
        expect(tokenizeDiceString("3d6+1+d8+1")).toMatchObject([
          [
            [3, 6],
            [1, 8],
          ],
          2,
        ]);
      });
    });

    describe("tokensToString()", () => {
      it("Converts tokens back to strings", () => {
        const _3d6plus1d8plus2 = [
          [
            [3, 6],
            [1, 8],
          ],
          2,
        ] as DiceTokensWithModifier;
        expect(tokensToString(_3d6plus1d8plus2)).toBe("3d6+1d8+2");
      });
    });

    describe("simplifyDiceString()", () => {
      it("Converts a dicestring into a cleaned up dice string", () => {
        expect(simplifyDiceString("2D6 +d8 + 4 -1+1d6 ")).toBe("3d6+1d8+3");
        expect(simplifyDiceString("d6+d6+d6")).toBe("3d6");
        expect(simplifyDiceString("1")).toBe("1");
        expect(simplifyDiceString("")).toBe("");
      });
      it("passes through errors unchanged", () => {
        expect(simplifyDiceString("3dd3")).toBe("3dd3");
      });
    });

    describe("getRange()", () => {
      it("Should take a set of tokens and return the maximum and minimum values.", () => {
        const _2d6 = [2, 6] as DiceToken;
        const _3d8 = [3, 8] as DiceToken;
        const _plus5 = 5 as Modifier;
        const tokens = [[_2d6, _3d8], _plus5] as DiceTokensWithModifier;

        const range = getRange(tokens);
        expect(range[0]).toBe(2 + 3 + 5);
        expect(range[1]).toBe(2 * 6 + 3 * 8 + 5);

        const tokens2 = [[[1, 6]], -6] as DiceTokensWithModifier;
        const range2 = getRange(tokens2);
        expect(range2[0]).toBe(-5);
        expect(range2[1]).toBe(0);

        const k = [[], 5] as DiceTokensWithModifier;
        const constantRange = getRange(k);
        expect(constantRange[0]).toBe(5);
        expect(constantRange[1]).toBe(5);
      });
      it("If the input value is a string instead of tokens, automatically converts to tokens.", () => {
        const range = getRange("2d6+3d8+5");
        expect(range[0]).toBe(2 + 3 + 5);
        expect(range[1]).toBe(2 * 6 + 3 * 8 + 5);
      });
    });
  });
});


src/utils.ts
import { NumberGenerator, Modifier } from "./types";

// identity :: a -> () -> a
// eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/explicit-module-boundary-types
export const identity =
  <T>(a: T) =>
  (_ignored?: any) =>
    a;

// add :: Number a => (a,a) -> a
export const add = (a: number, b: number): number => a + b;

// sum :: Number a => (...a[]) -> a
export const sum = (...numbers: number[]): number => numbers.reduce(add, 0);
// sumArray :: Number a => a[] -> a
export const sumArray = (numbers: number[]): number => sum(...numbers);

// max :: Number a => a[] -> a
export const max = (nums: number[]): number =>
  nums.reduce((a, b) => Math.max(a, b), Number.NEGATIVE_INFINITY);

// min :: Number a => a[] -> a
export const min = (nums: number[]): number =>
  nums.reduce((a, b) => Math.min(a, b), Number.POSITIVE_INFINITY);

// callMultipleTimes :: Number t => t -> (t -> a) -> a[]
export const callMultipleTimes =
  (times: number) =>
  (func: (index: number) => any): any[] => {
    let i = 0;
    const results = [];
    while (i < times) {
      results[i] = func(i);
      i += 1;
    }
    return results;
  };

// callAndAdd :: Number a => a -> (() -> a) -> (() -> a)
export const callAndAdd =
  (modifier: Modifier) =>
  (func: NumberGenerator): NumberGenerator =>
  () =>
    func() + modifier;

// count :: Number n => a -> a[] -> n
export const count =
  <T>(match: T) =>
  (list: T[]): number =>
    list.filter((element) => element === match).length;

// inclusive for min and max
export const randomIntegerBetween =
  (r: NumberGenerator) =>
  (possiblyMax: number, possiblyMin = 0): number => {
    // make sure max >= min
    const [trueMax, trueMin] = [
      Math.floor(Math.max(possiblyMax, possiblyMin)),
      Math.ceil(Math.min(possiblyMax, possiblyMin)),
    ];

    return Math.floor(r() * (trueMax - trueMin + 1) + trueMin);
  };

export const randomElement =
  (r: NumberGenerator) =>
  <T>(arr: T[]): T =>
    arr[randomIntegerBetween(r)(arr.length - 1)];

export const testRoll = callMultipleTimes;
export const testRollSm = testRoll(200);
export const testRollMed = testRoll(1000);
export const testRollLrg = testRoll(5000);
export const testRollXLrg = testRoll(100000);
